### PR TITLE
[BEAM-1265] Use state / timer API for DirectRunner timer firings

### PR DESCRIPTION
**Pull request by charlesccychen**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by charlesccychen** [comment]
R: @aaltay

**Comment by charlesccychen** [comment]
Python tests pass: https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/12524/org.apache.beam$beam-sdks-python/console ; the failure is a flake in the Java tests.

File: sdks/python/apache_beam/runners/direct/transform_evaluator.py [hunk 1]
```diff
@@ -199,6 +201,18 @@ def start_bundle(self):
...
+  def process_timer_wrapper(self, timer_firing):
+    """Process timer by clearing and then calling process_timer()."""
+    state = self.step_context.get_keyed_state(timer_firing.key)
+    state.clear_timer(
+        timer_firing.window, timer_firing.name, timer_firing.time_domain)
+    self.process_timer(timer_firing)
+
+  def process_timer(self, timer_firing):
+    """Default process_timer() impl. generating KeyedWorkItem element."""
+    self.process_element(
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/direct/transform_evaluator.py]: How does this work? Does `process_element` expect a `KeyedWorkItem` as an input?

File: sdks/python/apache_beam/runners/direct/transform_evaluator.py [hunk 2]
```diff
@@ -392,6 +405,9 @@ def start_bundle(self):
         self._applied_ptransform.transform.get_type_hints().input_types[0])
     self.key_coder = coders.registry.get_coder(kv_type_hint[0].tuple_types[0])
 
+  def process_timer(self, timer_firing):
+    pass
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/direct/transform_evaluator.py]: What is the reason for `pass` here?

File: sdks/python/apache_beam/runners/direct/transform_evaluator.py (hunk 1 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/transform_evaluator.py]: Thanks, I clarified this in the comments.

File: sdks/python/apache_beam/runners/direct/transform_evaluator.py (hunk 2 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/transform_evaluator.py]: Thanks, I clarified this in the comments.

**Comment by aaltay** [comment]
LGTM

**Pull request closed without merging**